curl-helper.c

```diff
@@ -1136,6 +1136,181 @@ static int cb_SSH_KEYFUNCTION(CURL *easy,
+#else
+  -1,
+#endif
+#if HAVE_DECL_CURLSSLBACKEND_BEARSSL
+  CURLSSLBACKEND_BEARSSL,
+#else
+  -1,
+#endif
+};
+
+#if HAVE_DECL_CURLSSLBACKEND_NONE
+typedef struct CURLsslsetMapping CURLsslsetMapping;
+
+struct CURLsslsetMapping
+{
+  char* name;
+  CURLsslset code;
+};
+
+/* Same order as in OCaml */
+CURLsslsetMapping sslsetMap[] = {
+  {"CURLSSLSET_OK", CURLSSLSET_OK},
+  {"CURLSSLSET_UNKNOWN_BACKEND", CURLSSLSET_UNKNOWN_BACKEND},
+  {"CURLSSLSET_TOO_LATE", CURLSSLSET_TOO_LATE},
+  {"CURLSSLSET_NO_BACKENDS", CURLSSLSET_NO_BACKENDS},
+};
+
+static void raiseSslsetError(CURLsslset err)
+{
+  CAMLparam0();
+  CAMLlocal1(exceptionData);
+  const value *exception;
+  const char *errorString = NULL;
+  int i;
+
+  for (i = 0; i < sizeof(sslsetMap) / sizeof(sslsetMap[0]); i ++) {
+    if (sslsetMap[i].code == err) {
+      errorString = sslsetMap[i].name;
+      break;
+    }
+  }
+
+  if (errorString == NULL)
+    caml_invalid_argument("Unexpected CURLsslset value");
+
+  exceptionData = caml_alloc_tuple(2);
+  Store_field(exceptionData, 0, Val_int(i));
+  Store_field(exceptionData, 1, caml_copy_string(errorString));
+
+  exception = caml_named_value("CurlSslSetException");
+  if (exception == NULL) caml_invalid_argument("CurlSslSetException not registered");
+
+  caml_raise_with_arg(*exception, exceptionData);
+
+  /* Not reached */
+  CAMLreturn0;
+}
+
+value caml_curl_global_sslset(value v_backend)
+{
+  CAMLparam1(v_backend);
+
+  curl_sslbackend backend = sslBackendMap[Int_val(v_backend)];
+  CURLsslset res = curl_global_sslset(backend, NULL, NULL);
+
+  if (res != CURLSSLSET_OK)
+    raiseSslsetError(res);
+
+  CAMLreturn(Val_unit);
+}
+
+value caml_curl_global_sslsetavail(value v_unit)
+{
+  CAMLparam1(v_unit);
+  CAMLlocal2(lst, v);
+  const curl_ssl_backend **backends;
+  CURLsslset res;
+  int i, j, n, found;
+
+  res = curl_global_sslset(-1, NULL, &backends);
+
+  if (res != CURLSSLSET_UNKNOWN_BACKEND)
+    raiseSslsetError(res);
+
+  for (n = 0; backends[n] != NULL; n ++) ;
+
+  lst = Val_emptylist;
+
+  for (i = n - 1; i >= 0; i --) {
+    found = -1;
+
+    for (j = 0; j < sizeof(sslBackendMap) / sizeof(sslBackendMap[0]); j ++) {
+      if (sslBackendMap[j] == backends[i]->id) {
+        found = j;
+        break;
+      }
+    }
+
+    /* If an unknown backend is returned, it is ignored */
+    if (found >= 0) {
+      v = caml_alloc_tuple(2);
+      Store_field(v, 0, Val_long(found));
+      Store_field(v, 1, lst);
+      lst = v;
+    }
+  }
+
+  CAMLreturn(lst);
+}
+#else
+value caml_curl_global_sslset(value v_backend)
+{
+  CAMLparam1(v_backend);
+  const value *exception = caml_named_value("Curl.NotImplemented");
+  if (NULL == exception) caml_invalid_argument("Curl.NotImplemented not registered");
+  caml_raise_with_string(*exception, "curl_global_sslset");
+
+  /* Not reached */
+  CAMLreturn(Val_unit);
+}
+value caml_curl_global_sslsetavail(value v_unit)
+{
+  return caml_curl_global_sslset(Val_int(0));
```

Comment:
Val_emptylist

Reply:
I don't know if it was clear from the code but here the argument is ignored, I changed the name of the arguments and added a comment to make this clear.